### PR TITLE
Emit event on Description-only healthcheck update

### PR DESCRIPTION
**Pull request by DamianSawicki**
Since https://github.com/kubernetes/ingress-gce/pull/2008, the health check description is modified if the service health check is configured with a BackendConfig CRD. The purpose of the present PR is to emit an event on the service on this occasion.

The changes are guarded with the flag `flags.F.EnableUpdateCustomHealthCheckDescription` from https://github.com/kubernetes/ingress-gce/pull/2018/.

### CODE REVIEW

File: pkg/healthchecks/healthchecks_test.go [hunk 1]
```diff
@@ -517,6 +520,11 @@ func TestRolloutUpdateCustomHCDescription(t *testing.T) {
 	if !reflect.DeepEqual(outputBCHC, outputBCHCWithFlag) {
 		t.Fatalf("Compute healthcheck is:\n%s, want:\n%s", pretty.Sprint(outputBCHC), pretty.Sprint(outputBCHCWithFlag))
 	}
+
+	output := <-fakeRecorder.Events
+	if !strings.HasPrefix(output, "Warning HealthcheckDescriptionUpdate") {
+		t.Fatalf("Incorrect event emitted on healthcheck update: %s.", output)
+	}
```
Comment (aojea) [inline pkg/healthchecks/healthchecks_test.go:527]: this can block forever, use select with another input time.After to fail the test if it timesout

File: vendor/k8s.io/cloud-provider-gcp/providers/gce/gce.go [hunk 2]
```diff
@@ -43,7 +43,7 @@ import (
 
 	"github.com/GoogleCloudPlatform/k8s-cloud-provider/pkg/cloud"
 
-	"k8s.io/api/core/v1"
+	v1 "k8s.io/api/core/v1"
```
Comment (aojea) [inline vendor/k8s.io/cloud-provider-gcp/providers/gce/gce.go:46]: vendor files should not be modified, maybe your IDE did this?

File: pkg/translator/healthchecks.go [hunk 3]
```diff
@@ -338,3 +341,32 @@ func ApplyProbeSettingsToHC(p *v1.Probe, hc *HealthCheck) {
 
 	hc.Description = DescriptionForHealthChecksFromReadinessProbe
 }
+
+// GetObjectKind implements runtime.Object
```
Comment (aojea) [inline pkg/translator/healthchecks.go:345]: is this change related?

File: pkg/healthchecks/healthchecks_test.go (hunk 1 above)
Comment (DamianSawicki) [reply pkg/healthchecks/healthchecks_test.go:527]: I tried this (with modified settings or in another test) and the test timed out after 30 seconds.

File: pkg/healthchecks/healthchecks.go [hunk 4]
```diff
@@ -142,6 +144,11 @@ func (h *HealthChecks) sync(hc *translator.HealthCheck, bchcc *backendconfigv1.H
 	changes := calculateDiff(filter(existingHC), filter(hc), bchcc)
 	if changes.hasDiff() {
 		klog.V(2).Infof("Health check %q needs update (%s)", existingHC.Name, changes)
+		if flags.F.EnableUpdateCustomHealthCheckDescription && changes.size() == 1 && changes.has("Description") {
+			message := fmt.Sprintf("Healthcheck will be updated and the only field updated is Description.\nOld: %+v\nNew: %+v\nDiff: %+v", existingHC, hc, changes)
```
Comment (aojea) [inline pkg/healthchecks/healthchecks.go:148]: this seems a very large message from an event, can we just send the diff or do we miss some important information?

**Comment by aojea** [comment]
/ok-to-test

looks good overall, some comments that needs to be addressed

File: vendor/k8s.io/cloud-provider-gcp/providers/gce/gce.go (hunk 2 above)
Comment (DamianSawicki) [reply vendor/k8s.io/cloud-provider-gcp/providers/gce/gce.go]: Right, it was my IDE. Thanks!

File: pkg/translator/healthchecks.go (hunk 3 above)
Comment (DamianSawicki) [reply pkg/translator/healthchecks.go:345]: Yes. `EventRecorder.Event` requires the first argument to be `runtime.Object`, and this interface contains two methods: `GetObjectKind() schema.ObjectKind` and `DeepCopyObject() Object`.

File: pkg/healthchecks/healthchecks_test.go (hunk 1 above)
Comment (DamianSawicki) [reply pkg/healthchecks/healthchecks_test.go]: Done.

File: pkg/healthchecks/healthchecks.go (hunk 4 above)
Comment (aojea) [reply pkg/healthchecks/healthchecks.go:148]: the message contains the New and Old too, can we send only the changes? do we need the others?

```
%+v\nNew: %+v\nDiff: %+v", existingHC, hc, changes)
```

File: pkg/healthchecks/healthchecks_test.go [hunk 5]
```diff
@@ -517,6 +521,16 @@ func TestRolloutUpdateCustomHCDescription(t *testing.T) {
 	if !reflect.DeepEqual(outputBCHC, outputBCHCWithFlag) {
 		t.Fatalf("Compute healthcheck is:\n%s, want:\n%s", pretty.Sprint(outputBCHC), pretty.Sprint(outputBCHCWithFlag))
 	}
+
+	select {
+	case output := <-fakeRecorder.Events:
+		if !strings.HasPrefix(output, "Warning HealthcheckDescriptionUpdate") {
+			t.Fatalf("Incorrect event emitted on healthcheck update: %s.", output)
+		}
+	case <-time.After(2 * time.Second):
```
Comment (aojea) [inline pkg/healthchecks/healthchecks_test.go:530]: give it some room, CIs can be very slow, 10 seconds is ok, we just want to fail and now why in case it is blocked, updated the t.Fatalf to indicate that we timed out

File: pkg/translator/healthchecks.go (hunk 3 above)
Comment (aojea) [reply pkg/translator/healthchecks.go:345]: the object in this case is the Service or just pass the Namespace, doing this to the healtcheck object seems very brittle

Comment (aojea) [reply pkg/translator/healthchecks.go:345]: I think you can get it from `defaultBackendSvc types.NamespacedName`

File: pkg/healthchecks/healthchecks.go (hunk 4 above)
Comment (DamianSawicki) [reply pkg/healthchecks/healthchecks.go:148]: Yes, I'm trying to say that we need the others. The `changes` aka "diff" inside the `if` consists only of the modified Description (`changes.size() == 1 && changes.has("Description")`). However, we want to include more information in the message as a precaution.

A background fact that I should probably mention is that `calculateDiff()` does not return the full diff. It checks selected fields only (recently `Description` was added as one of them), sometimes just 2 of them.

File: pkg/translator/healthchecks.go (hunk 3 above)
Comment (DamianSawicki) [reply pkg/translator/healthchecks.go:345]: This again requires implementing the above methods either for `string` (in the case of namespace) or for `types.NamespacedName` (in the case of service), and when trying to do that, I'm getting the following error `cannot define new methods on non-local type "k8s.io/apimachinery/pkg/types".NamespacedName`. I guess I can write my own alias type for one of these and then implement the methods for the alias. @aojea, do you think it is a good idea?

File: pkg/healthchecks/healthchecks_test.go (hunk 5 above)
Comment (DamianSawicki) [reply pkg/healthchecks/healthchecks_test.go]: Done.

File: pkg/translator/healthchecks.go (hunk 3 above)
Comment (aojea) [reply pkg/translator/healthchecks.go:345]: no sorry, my point is that we should not modify this type, and use one of the Kubernetes types associated to the healtcheck, it can be the Service or the Namespace, ideally the Service

Comment (DamianSawicki) [reply pkg/translator/healthchecks.go:347]: Thanks for the clarification! Done, please have a look.

Comment (aojea) [reply pkg/translator/healthchecks.go:347]: you forgot to remove this code

File: pkg/healthchecks/healthchecks.go [hunk 6]
```diff
@@ -142,6 +160,11 @@ func (h *HealthChecks) sync(hc *translator.HealthCheck, bchcc *backendconfigv1.H
...
+			h.eventRecorder.Event(
+				hc.Service, "Warning", "HealthcheckDescriptionUpdate", message)
```
Comment (aojea) [inline pkg/healthchecks/healthchecks.go:166]: hmm, I have to apologize, but now that I look it carefully , it seems Service is not the right object to pass, in this case this method is trying to sync `bchcc *backendconfigv1.HealthCheckConfig` , I think this object is implementing `runtime.Object`, isnt't it?, that will be the best thing since we can do this only with 4-5 lines of code

Sorry again, I've should have checked this better before doing any suggestion

File: pkg/backends/backends.go [hunk 7]
```diff
@@ -41,7 +42,7 @@ type Backends struct {
 }
 
 // Backends is a Pool.
-var _ Pool = (*Backends)(nil)
+var _ interfaces.Pool = (*Backends)(nil)
```
Comment (kl52752) [inline pkg/backends/backends.go:45]: why this change was needed?

Comment (kl52752) [reply pkg/backends/backends.go:45]: I think you need to rebase

File: pkg/healthchecks/healthchecks.go (hunk 6 above)
Comment (kl52752) [reply pkg/healthchecks/healthchecks.go:166]: @aojea we want to emit the even on the Service because UHC is created per Service and one BackendConfig can be assign to multiple Services but all Customer cares is a configuration of HealthCheck for Service.

File: pkg/backends/interfaces/interfaces.go [hunk 8]
```diff
@@ -77,6 +77,21 @@ type NEGGetter interface {
 	GetNetworkEndpointGroup(name string, zone string, version meta.Version) (*composite.NetworkEndpointGroup, error)
 }
 
+// ServiceGetter is an interface to retrieve Kubernetes Services.
+type ServiceGetter interface {
+	GetService(namespace, name string) (*api_v1.Service, error)
+}
+
+func NewEmptyServiceGetter() ServiceGetter {
```
Comment (kl52752) [inline pkg/backends/interfaces/interfaces.go:85]: NewFake

File: pkg/backends/interfaces/interfaces.go [hunk 9]
```diff
@@ -77,6 +77,21 @@ type NEGGetter interface {
...
+	return &fakeServiceGetter{}
+}
+
+type fakeServiceGetter struct{}
+
+func (fsg *fakeServiceGetter) GetService(namespace, name string) (*api_v1.Service, error) {
```
Comment (kl52752) [inline pkg/backends/interfaces/interfaces.go:91]: I think it will be better to return some default Service here. I don't think is good to return the empty object. Normally when you get nil error you expect that the Object is properly constructed. This may lead to some crashes I think.

File: pkg/controller/controller.go [hunk 10]
```diff
@@ -114,7 +115,8 @@ func NewLoadBalancerController(
 		Interface: ctx.KubeClient.CoreV1().Events(""),
 	})
 
-	healthChecker := healthchecks.NewHealthChecker(ctx.Cloud, ctx.HealthCheckPath, ctx.DefaultBackendSvcPort.ID.Service)
+	eventRecorder := ctx.Recorder(ctx.Namespace)
```
Comment (kl52752) [inline pkg/controller/controller.go:118]: the namespace for recorder needs to be the namespace of the Service you wants to emit the event for. And health checker is created one per controller you need to postpone this function to the place where you have your Service. 
You probably need context there.

File: pkg/backends/interfaces/interfaces.go (hunk 9 above)
Comment (DamianSawicki) [reply pkg/backends/interfaces/interfaces.go:91]: Actually, according to the comments ,`api_v1.Service` has only one non-optional field of type `metav1.TypeMeta`, which in turn has only optional fields. So it seems that the "empty object" is a valid object.

Comment (kl52752) [reply pkg/backends/interfaces/interfaces.go:91]: you can at least fill name and namespace to match function params.
Where is it used?

File: pkg/healthchecks/healthchecks.go (hunk 6 above)
Comment (aojea) [reply pkg/healthchecks/healthchecks.go:166]: yeah, Damien commented offline, sounds good

Comment (aojea) [reply pkg/healthchecks/healthchecks.go:166]: yeah, Damien commented offline, sounds good

File: pkg/backends/interfaces/interfaces.go (hunk 9 above)
Comment (aojea) [reply pkg/backends/interfaces/interfaces.go:91]: this should be enough
```
	svc := &v1.Service{
		ObjectMeta: metav1.ObjectMeta{
			Name:      name,
			Namespace: namespace,
		},
	}
```

**Comment by aojea** [comment]
just please don't use calls to the apiserver on the getter, remember that you have informers that already have Listers and Getter to a local cache with all the objects information. The other alternative is to plumb down the information on initialization, I'm not familiar with the code, so I can't suggest which one is the best option

File: pkg/controller/controller.go (hunk 10 above)
Comment (DamianSawicki) [reply pkg/controller/controller.go:118]: Done.

File: pkg/backends/interfaces/interfaces.go (hunk 9 above)
Comment (DamianSawicki) [reply pkg/backends/interfaces/interfaces.go:91]: It's just used as one of the Event parameters.

File: pkg/backends/interfaces/interfaces.go (hunk 8 above)
Comment (DamianSawicki) [reply pkg/backends/interfaces/interfaces.go:85]: Done.

File: pkg/backends/backends.go (hunk 7 above)
Comment (DamianSawicki) [reply pkg/backends/backends.go:45]: Discussed offline and fixed.

File: pkg/translator/healthchecks.go (hunk 3 above)
Comment (DamianSawicki) [reply pkg/translator/healthchecks.go:347]: Done.

File: pkg/backends/interfaces/interfaces.go (hunk 9 above)
Comment (DamianSawicki) [reply pkg/backends/interfaces/interfaces.go]: @aojea Yeah, I missed your comment earlier, but this is exactly what we ended up with.

File: pkg/healthchecks/healthchecks.go [hunk 11]
```diff
@@ -67,9 +69,23 @@ func (h *HealthChecks) new(sp utils.ServicePort) *translator.HealthCheck {
 	hc.Name = sp.BackendName()
 	hc.Port = sp.NodePort
 	hc.RequestPath = h.pathFromSvcPort(sp)
+	service := h.mainService(sp)
```
Comment (DamianSawicki) [inline pkg/healthchecks/healthchecks.go:72]: I'll flag-gate the following lines.

Comment (DamianSawicki) [reply pkg/healthchecks/healthchecks.go]: Done.

File: pkg/healthchecks/healthchecks.go [hunk 12]
```diff
@@ -142,6 +160,11 @@ func (h *HealthChecks) sync(hc *translator.HealthCheck, bchcc *backendconfigv1.H
 	changes := calculateDiff(filter(existingHC), filter(hc), bchcc)
 	if changes.hasDiff() {
 		klog.V(2).Infof("Health check %q needs update (%s)", existingHC.Name, changes)
+		if flags.F.EnableUpdateCustomHealthCheckDescription && changes.size() == 1 && changes.has("Description") {
+			message := fmt.Sprintf("Healthcheck will be updated and the only field updated is Description.\nOld: %+v\nNew: %+v\nDiff: %+v", existingHC, hc, changes)
+			h.recorderGetter.Recorder(hc.Service.Namespace).Event(
```
Comment (cezarygerard) [inline pkg/healthchecks/healthchecks.go:165]: hc.Service Can be null here,

if it error occurred while assigning the hc.Service in pkg/healthchecks/healthchecks.go
```
if flags.F.EnableUpdateCustomHealthCheckDescription {
		service := h.mainService(sp)
		var err error
		hc.Service, err = h.serviceGetter.GetService(service.Namespace, service.Name)
		if err != nil {
			klog.Errorf("Service %s/%s not found: %v.", service.Namespace, service.Name, err)
		}
	}
```

File: pkg/healthchecks/healthchecks.go [hunk 13]
```diff
@@ -67,9 +69,25 @@ func (h *HealthChecks) new(sp utils.ServicePort) *translator.HealthCheck {
 	hc.Name = sp.BackendName()
 	hc.Port = sp.NodePort
 	hc.RequestPath = h.pathFromSvcPort(sp)
+	if flags.F.EnableUpdateCustomHealthCheckDescription {
```
Comment (cezarygerard) [inline pkg/healthchecks/healthchecks.go:72]: you should not be using flags this deep down the logic

convert them into fields somewhere in context or close to main

it makes it much more testable

File: pkg/healthchecks/healthchecks.go [hunk 14]
```diff
@@ -67,9 +69,25 @@ func (h *HealthChecks) new(sp utils.ServicePort) *translator.HealthCheck {
...
+		service := h.mainService(sp)
+		var err error
+		hc.Service, err = h.serviceGetter.GetService(service.Namespace, service.Name)
+		if err != nil {
+			klog.Errorf("Service %s/%s not found: %v.", service.Namespace, service.Name, err)
```
Comment (cezarygerard) [inline pkg/healthchecks/healthchecks.go:77]: 1. why erorr?
we keep going so it should be a Warning

2. Add one more sentence to this log, what are the consequences (not big, we keep going)

File: pkg/healthchecks/healthchecks.go [hunk 15]
```diff
@@ -67,9 +69,25 @@ func (h *HealthChecks) new(sp utils.ServicePort) *translator.HealthCheck {
 	hc.Name = sp.BackendName()
 	hc.Port = sp.NodePort
 	hc.RequestPath = h.pathFromSvcPort(sp)
+	if flags.F.EnableUpdateCustomHealthCheckDescription {
+		service := h.mainService(sp)
```
Comment (cezarygerard) [inline pkg/healthchecks/healthchecks.go:73]: service is a confusing name here
It suggests it's v1.Service
but it is just name and namespace

see my next comment

File: pkg/healthchecks/healthchecks.go [hunk 16]
```diff
@@ -67,9 +69,25 @@ func (h *HealthChecks) new(sp utils.ServicePort) *translator.HealthCheck {
...
+		var err error
+		hc.Service, err = h.serviceGetter.GetService(service.Namespace, service.Name)
+		if err != nil {
+			klog.Errorf("Service %s/%s not found: %v.", service.Namespace, service.Name, err)
+		}
+	}
```
Comment (cezarygerard) [inline pkg/healthchecks/healthchecks.go:79]: this whole part can be well hidden:
see this:
```
// new returns a *HealthCheck with default settings and specified port/protocol
func (h *HealthChecks) new(sp utils.ServicePort) *translator.HealthCheck {
	var hc *translator.HealthCheck
	if sp.NEGEnabled && !sp.L7ILBEnabled {
		hc = translator.DefaultNEGHealthCheck(sp.Protocol)
	} else if sp.L7ILBEnabled {
		hc = translator.DefaultILBHealthCheck(sp.Protocol)
	} else {
		hc = translator.DefaultHealthCheck(sp.NodePort, sp.Protocol)
	}
	// port is the key for retrieving existing health-check
	// TODO: rename backend-service and health-check to not use port as key
	hc.Name = sp.BackendName()
	hc.Port = sp.NodePort
	hc.RequestPath = h.pathFromSvcPort(sp)
	hc.Service = h.getService(sp)

	return hc
}

func (h *HealthChecks) getService(sp utils.ServicePort) *v1.Service {
	if flags.F.EnableUpdateCustomHealthCheckDescription {
		namespacedName := h.mainService(sp)
		var err error
		service, err := h.serviceGetter.GetService(namespacedName.Namespace, namespacedName.Name)
		if err != nil {
			klog.Errorf("Service %s/%s not found: %v.", namespacedName.Namespace, namespacedName.Name, err)
		}
		return service
	}
	return nil
}

func (h *HealthChecks) mainService(sp utils.ServicePort) types.NamespacedName {
	service := h.defaultBackendSvc
	if sp.ID.Service.Name != "" {
		service = sp.ID.Service
	}
	return service
}
```

File: pkg/healthchecks/healthchecks.go [hunk 17]
```diff
@@ -142,6 +160,11 @@ func (h *HealthChecks) sync(hc *translator.HealthCheck, bchcc *backendconfigv1.H
 	changes := calculateDiff(filter(existingHC), filter(hc), bchcc)
 	if changes.hasDiff() {
 		klog.V(2).Infof("Health check %q needs update (%s)", existingHC.Name, changes)
+		if flags.F.EnableUpdateCustomHealthCheckDescription && changes.size() == 1 && changes.has("Description") {
+			message := fmt.Sprintf("Healthcheck will be updated and the only field updated is Description.\nOld: %+v\nNew: %+v\nDiff: %+v", existingHC, hc, changes)
+			h.recorderGetter.Recorder(hc.Service.Namespace).Event(
+				hc.Service, "Warning", "HealthcheckDescriptionUpdate", message)
```
Comment (cezarygerard) [inline pkg/healthchecks/healthchecks.go:166]: this is Info type of message, not Warning

be cautious what we log to customer projects ;-)

Comment (cezarygerard) [reply pkg/healthchecks/healthchecks.go:166]: also use const from this package
vendor/k8s.io/api/core/v1/types.go

as we do in other places
```
// Valid values for event types (new types could be added in future)
const (
	// Information only and will not cause any problems
	EventTypeNormal string = "Normal"
	// These events are to warn that something might go wrong
	EventTypeWarning string = "Warning"
)
```

File: pkg/healthchecks/interfaces.go [hunk 18]
```diff
@@ -56,3 +58,57 @@ type HealthChecker interface {
...
+func (sfrg *singletonFakeRecorderGetter) FakeRecorder() *record.FakeRecorder {
+	if sfrg.recorder == nil {
+		panic("singletonFakeRecorderGetter not initialised: recorder is nil.")
+	}
+	return sfrg.recorder
+}
+
+func NewFakeSingletonRecorderGetter(bufferSize int) *singletonFakeRecorderGetter {
+	return &singletonFakeRecorderGetter{recorder: record.NewFakeRecorder(bufferSize)}
+}
```
Comment (cezarygerard) [inline pkg/healthchecks/interfaces.go:114]: all fakes should be moved to _test.go file

File: pkg/healthchecks/healthchecks.go (hunk 12 above)
Comment (DamianSawicki) [reply pkg/healthchecks/healthchecks.go:165]: Great catch!

File: pkg/healthchecks/healthchecks.go (hunk 16 above)
Comment (DamianSawicki) [reply pkg/healthchecks/healthchecks.go:79]: Much better, I somehow missed it!

File: pkg/healthchecks/healthchecks.go (hunk 17 above)
Comment (DamianSawicki) [reply pkg/healthchecks/healthchecks.go]: Done.

File: pkg/healthchecks/healthchecks.go (hunk 16 above)
Comment (DamianSawicki) [reply pkg/healthchecks/healthchecks.go]: Done.

File: pkg/healthchecks/healthchecks.go (hunk 15 above)
Comment (DamianSawicki) [reply pkg/healthchecks/healthchecks.go]: Done.

File: pkg/healthchecks/healthchecks.go (hunk 14 above)
Comment (DamianSawicki) [reply pkg/healthchecks/healthchecks.go]: Done.

File: pkg/healthchecks/healthchecks.go (hunk 12 above)
Comment (DamianSawicki) [reply pkg/healthchecks/healthchecks.go]: Done.

File: pkg/healthchecks/healthchecks.go (hunk 13 above)
Comment (DamianSawicki) [reply pkg/healthchecks/healthchecks.go]: We had this discussion already with kl52752, panslava, and bowei: https://github.com/kubernetes/ingress-gce/pull/2018#discussion_r1145665678. This is the same flag as there and we're not adding any new non-parallelizable tests in this PR, so I'll stick to the approach bowei suggested in https://github.com/kubernetes/ingress-gce/pull/2018 and keep the flag deep down.

File: pkg/healthchecks/interfaces.go (hunk 18 above)
Comment (DamianSawicki) [reply pkg/healthchecks/interfaces.go]: Moved to a new `flakes.go` file because somehow when I moved it to `healthchecks_test.go` it was not visible in other packages.

Comment (cezarygerard) [reply pkg/healthchecks/interfaces.go]: yeah... so it has to be compiled into final binary

I have tried to use build constraint 
```
//go:build test
```
or 
```
//go:build testing
```
but it does not work


golang keeps amazing me :-D

**Review by cezarygerard — APPROVED** [review]

File: pkg/healthchecks/interfaces.go (hunk 18 above)
Comment (DamianSawicki) [reply pkg/healthchecks/interfaces.go]: https://stackoverflow.com/a/22277807
